catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswSubscriptionEndpoint.java

```diff
@@ -0,0 +1,546 @@
+        // to registry
+        if (registeredSubscriptions.containsKey(subscriptionUuid)) {
+            LOGGER.debug("Delete existing subscription {} for re-creation", subscriptionUuid);
+            deleteSubscription(subscriptionUuid);
+        }
+        CswSubscription sub = createSubscription(request);
+        LOGGER.debug("Registering Subscription");
+        ServiceRegistration serviceRegistration =
+                context.registerService(Subscription.class.getName(), sub, props);
+
+        if (serviceRegistration != null) {
+            LOGGER.debug("Subscription registered with bundle ID = {} ",
+                    serviceRegistration.getReference()
+                            .getBundle()
+                            .getBundleId());
+            registeredSubscriptions.put(subscriptionUuid, serviceRegistration);
+            // Pass in client-provided subscriptionId vs. subscription UUID because
+            // the filter XML to be persisted for this subscription will be used to
+            // restore this subscription and should consist of the exact values the
+            // client originally provided.
+            if (persistSubscription) {
+                persistSubscription(sub, deliveryMethodUrl, subscriptionUuid);
+            }
+        } else {
+            LOGGER.debug("Subscription registration failed");
+        }
+
+        LOGGER.debug("EXITING: {}", methodName);
+        return subscriptionUuid;
+
+    }
+
+    private synchronized boolean deleteSubscription(String subscriptionId) throws CswException {
+        String methodName = "deleteSubscription";
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("ENTERING: {}", methodName);
+            LOGGER.debug("subscriptionId = {}", subscriptionId);
+        }
+
+        if (StringUtils.isEmpty(subscriptionId)) {
+            throw new CswException(
+                    "Unable to delete subscription because subscription ID is null or empty");
+        }
+
+        boolean status = false;
+
+        try {
+            LOGGER.debug("Removing (unregistering) subscription: {}", subscriptionId);
+            ServiceRegistration sr = (ServiceRegistration) registeredSubscriptions.remove(
+                    subscriptionId);
+            if (sr != null) {
+                sr.unregister();
+            } else {
+                LOGGER.debug("No ServiceRegistration found for subscription: {}", subscriptionId);
+            }
+
+            Configuration subscriptionConfig = getSubscriptionConfiguration(subscriptionId);
+            try {
+                if (subscriptionConfig != null) {
+                    LOGGER.debug("Deleting subscription for subscriptionId = {}", subscriptionId);
+                    subscriptionConfig.delete();
+
+                    // Subscription removal is only successful if able to remove from OSGi registry and
+                    // ConfigAdmin service
+                    if (sr != null) {
```

Comment:
So what happens if we could remove the subscription from the map but couldn't delete it from config admin and they end up being out-of-sync? Should we invalidate the map at that point since it looks like we're really using it as a cache? Also, where do we initialize the map with the Config Admin data when this class gets created?

Really feels like separating the data store (config admin) code from the caching (registeredSubscriptions map) code from the rest of the business logic might be a good idea here. Especially if we eventually want to move away from using the Config Admin to persist this information and use something like a key-value storage solution such as the `PersistentStore`.